File: 04. LinkedList/06. Swap nodes in pairs/main.java
class ListNode {
    int val;
    ListNode next;
    
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class main {
    public static ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode newHead = head.next;
        ListNode prev = null;

        while (head != null && head.next != null) {
            ListNode first = head;
            ListNode second = head.next;

            first.next = second.next;
            second.next = first;

            if (prev != null) prev.next = second;
            prev = first;
            head = first.next;
        }

        return newHead;
    }

    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("NULL");
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);

        System.out.print("Original List: ");
        printList(head);

        head = swapPairs(head);

        System.out.print("List after swapping pairs: ");
        printList(head);
    }
}

File: 04. LinkedList/06. Swap nodes in pairs/main.js
class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

function swapPairs(head) {
    if (!head || !head.next) return head;

    let newHead = head.next;
    let prev = null;

    while (head && head.next) {
        let first = head;
        let second = head.next;

        first.next = second.next;
        second.next = first;

        if (prev) prev.next = second;
        prev = first;
        head = first.next;
    }

    return newHead;
}

// Utility function to print a linked list
function printList(head) {
    let result = "";
    while (head) {
        result += head.val + " -> ";
        head = head.next;
    }
    result += "NULL";
    console.log(result);
}

// Main function
function main() {
    let head = new ListNode(1);
    head.next = new ListNode(2);
    head.next.next = new ListNode(3);
    head.next.next.next = new ListNode(4);

    console.log("Original List:");
    printList(head);

    head = swapPairs(head);

    console.log("List after swapping pairs:");
    printList(head);
}

// Run main function
main();
